txscript/opcode.go

```diff
@@ -1088,6 +1100,76 @@ func opcodeCheckLockTimeVerify(op *parsedOpcode, vm *Engine) error {
 	return nil
 }
 
+// opcodeCheckSequenceVerify compares the top item on the data stack to the
+// LockTime field of the transaction containing the script signature
+// validating if the transaction outputs are spendable yet.  If flag
+// ScriptVerifyCheckSequenceVerify is not set, the code continues as if OP_NOP3
+// were executed.
+func opcodeCheckSequenceVerify(op *parsedOpcode, vm *Engine) error {
+	// If the ScriptVerifyCheckSequenceVerify script flag is not set, treat
+	// opcode as OP_NOP3 instead.
+	if !vm.hasFlag(ScriptVerifyCheckSequenceVerify) {
+		if vm.hasFlag(ScriptDiscourageUpgradableNops) {
+			return errors.New("OP_NOP3 reserved for soft-fork " +
+				"upgrades")
+		}
+		return nil
+	}
+
+	if vm.tx.Version < 2 {
+		return fmt.Errorf("invalid transaction version: %d",
+			vm.tx.Version)
+	}
+
+	// PeekByteArray is used here instead of PeekInt because we do not want
+	// to be limited to a 4-byte integer for reasons specified above.
+	so, err := vm.dstack.PeekByteArray(0)
+	if err != nil {
+		return err
+	}
+
+	stackSequence, err := makeScriptNum(so, vm.dstack.verifyMinimalData, 5)
+	if err != nil {
+		return err
+	}
+
+	// In the rare event that the argument may be < 0 due to some arithmetic
+	// being done first, you can always use 0 OP_MAX OP_CHECKSEQUENCEVERIFY.
+	if stackSequence < 0 {
+		return fmt.Errorf("negative sequence: %d", stackSequence)
+	}
+
+	sequence := int64(stackSequence)
+
+	// To provide for future soft-fork extensibility, if the
+	// operand has the disabled lock-time flag set,
+	// CHECKSEQUENCEVERIFY behaves as a NOP.
+	if sequence&int64(wire.SequenceLockTimeDisabled) != 0 {
+		return nil
+	}
+
+	// Sequence numbers with their most significant bit set are not
+	// consensus constrained. Testing that the transaction's sequence
+	// number do not have this bit set prevents using this property
+	// to get around a CHECKSEQUENCEVERIFY check.
+	txSequence := int64(vm.tx.TxIn[vm.txIdx].Sequence)
+	if txSequence&int64(wire.SequenceLockTimeDisabled) != 0 {
+		return fmt.Errorf("transaction sequence has sequence "+
+			"locktime disabled bit enabled: 0x%x", txSequence)
```

Comment:
I'd suggest `disabled bit set` as opposed to `disabled bit enabled` as it reads a little more clearly in my opinion.